subsys/bluetooth/shell/bt.c

```diff
@@ -193,65 +193,84 @@ static bool data_cb(struct bt_data *data, void *user_data)
 	}
 }
 
-static void print_data_set(uint8_t type, uint8_t set_value_len,
-			   const uint8_t *scan_data, uint8_t scan_data_len)
+static void print_data_hex(const uint8_t *data, uint8_t len)
 {
-	uint8_t min_value_len = MIN(set_value_len, scan_data_len);
+	shell_fprintf(ctx_shell, SHELL_INFO, "0x");
+	for (int8_t i = len-1; i >= 0; i--) {
+		shell_fprintf(ctx_shell, SHELL_INFO, "%02x", data[i]);
+	}
+}
 
-	shell_fprintf(ctx_shell, SHELL_INFO, "%*sType 0x%02x: ",
-		      strlen(scan_response_label), "",
-		      type);
-	for (uint8_t i = 0U; i < (scan_data_len / min_value_len); i++) {
-		if (i > 0L) {
-			shell_fprintf(ctx_shell, SHELL_INFO, ", ");
-		}
+static void print_data_set(uint8_t set_value_len, bool single_value,
+			   const uint8_t *scan_data, uint8_t scan_data_len)
+{
+	uint8_t idx = 0;
 
-		shell_fprintf(ctx_shell, SHELL_INFO, "0x");
-		for (uint8_t j = 0U; j < min_value_len; j++) {
-			shell_fprintf(ctx_shell, SHELL_INFO, "%02x",
-				      scan_data[i * min_value_len + j]);
+	do {
+		if (idx > 0) {
+			shell_fprintf(ctx_shell, SHELL_INFO, ",");
 		}
-	}
-
-	shell_fprintf(ctx_shell, SHELL_INFO, "\n");
+		print_data_hex(&scan_data[idx], set_value_len);
+		idx += set_value_len;
+		if (single_value)
+			break;
```

Comment:
Is this necessary? 

Is it not possible to check if `idx == scan_data_len` or provide the right arguments so that the check in the `while` would stop after the first if that is what the caller wants?

Reply:
I will put the check for single_value in the while()

Reply:
My point is that the `single_value` boolean shouldn't be necessary. See https://github.com/zephyrproject-rtos/zephyr/pull/58937#discussion_r1224188190